File: java/src/main/java/org/sammancoaching/StagingDeployStep.java
package org.sammancoaching;

import org.sammancoaching.dependencies.Config;
import org.sammancoaching.dependencies.DeploymentEnvironment;
import org.sammancoaching.dependencies.Logger;
import org.sammancoaching.dependencies.Project;

public record StagingDeployStep(String name, Config config, Logger log) implements PipelineStep {
    @Override
    public PipelineStepResult run(Project project, PipelineStepResult previousStepResult) {
        boolean stepPassed = false;
        String failureReason = previousStepResult.failureReason();
        if (previousStepResult.stepPassed()) {
            if ("success".equals(project.deploy(DeploymentEnvironment.STAGING))) {
                log.info(name() + " successful");
                stepPassed = true;
            } else {
                String reason = name() + " failed";
                log.error(reason);
                failureReason = reason;
            }
        }
        return new PipelineStepResult(this.name(), stepPassed, failureReason);
    }
}

File: csharp/BuildPipeline/StagingDeployStep.cs
namespace UntangledConditionals
{
    public class StagingDeployStep : PipelineStep
    {
        private readonly string _name;
        private readonly Config _config;
        private readonly Logger _log;

        public StagingDeployStep(string name, Config config, Logger log)
        {
            _name = name;
            _config = config;
            _log = log;
        }

        public PipelineStepResult Run(Project project, PipelineStepResult previousStepResult)
        {
            bool stepPassed = false;
            string failureReason = previousStepResult.FailureReason;

            if (previousStepResult.StepPassed)
            {
                if ("success".Equals(project.Deploy(DeploymentEnvironment.Staging)))
                {
                    _log.Info(_name + " successful");
                    stepPassed = true;
                }
                else
                {
                    string reason = _name + " failed";
                    _log.Error(reason);
                    failureReason = reason;
                }
            }

            return new PipelineStepResult(_name, stepPassed, failureReason);
        }
    }
}

